lib/puppet/pops/types/implementation_registry.rb

```diff
@@ -0,0 +1,143 @@
+module Puppet::Pops
+module Types
+  # The {ImplementationRegistry} maps names types in the Puppet Type System to names of corresponding implementation
+  # modules/classes. Each mapping is unique and bidirectional so that for any given type name there is only one
+  # implementation and vice versa.
+  #
+  # @api private
+  class ImplementationRegistry
+    NAME_SEGMENT_SEPARATOR =
+    TYPE_REGEXP_SUBST = TypeFactory.tuple([PRegexpType::DEFAULT, PStringType::NON_EMPTY])
+
+    def self.singleton
+      @singleton ||= new(Loaders.static_loader)
+    end
+
+    # Create a new instance. This method is normally only called once
+    #
+    # The initializer will create mappings for well known types that can be loaded using the static loader
+    #
+    def initialize(static_loader)
+      @type_names_per_implementation = {}
+      @implementations_per_type_name = {}
+      @type_name_substitutions = []
+      @impl_name_substitutions = []
+
+      register_implementation('PCore::AST::Locator', 'Puppet::Pops::Parser::Locator::Locator19', static_loader)
+      register_implementation_namespace('PCore::AST', 'Puppet::Pops::Model', static_loader)
+
+      # Register all known type implementations. The Type itself does not follow the standard naming convention
+      register_implementation('Type', 'Puppet::Pops::Types::PType', static_loader)
+
+      # The rest of the types all follow a standard naming convention
+      register_implementation_regexp([/^([A-Z]\w*)$/, 'Puppet::Pops::Types::P\1Type'], [/^Puppet::Pops::Types::P(\w+)Type$/, '\1'], static_loader)
+    end
+
+    # Register a bidirectional type mapping.
+    #
+    # @override register_type_mapping(runtime_type, puppet_type)
```

Comment:
You probably meant "@overload" (IIRC, "@override" is proposed but not implemented yet in yard + this is not an override).